hoomd/mpcd/pytest/test_snapshot.py

```diff
@@ -0,0 +1,261 @@
+        snap.mpcd.typeid[:] = typeid0
+        snap.mpcd.types = types0
+
+    snap.replicate(nx=nx, ny=ny, nz=nz)
+
+    if snap.communicator.rank == 0:
+        # replicate the numpy position array along x, y, and z
+        pos1 = []
+        vel1 = []
+        typeid1 = []
+        new_origin = L * np.array([nx / 2, ny / 2, nz / 2])
+        for i in range(0, nx):
+            for j in range(0, ny):
+                for k in range(0, nz):
+                    for r, v, tid in zip(pos0, vel0, typeid0):
+                        dr = np.array([i * L, j * L, k * L])
+                        pos1 += [list(r + dr + 0.5 * L - new_origin)]
+                        vel1 += [v]
+                        typeid1 += [tid]
+
+        scale = nx * ny * nz
+        assert snap.mpcd.N == len(pos0) * scale
+        np.testing.assert_allclose(snap.mpcd.position, pos1)
+        np.testing.assert_allclose(snap.mpcd.velocity, vel1)
+        np.testing.assert_array_equal(snap.mpcd.typeid, typeid1)
+        np.testing.assert_equal(snap.mpcd.types, types0)
+
+
+def test_create_from_snap(snap, simulation_factory):
+    # set snap values and initialize
+    if snap.communicator.rank == 0:
+        snap.mpcd.N = test_positions.shape[0]
+        snap.mpcd.position[:] = test_positions
+        snap.mpcd.velocity[:] = test_velocities
+        snap.mpcd.typeid[:] = test_typeids
+        snap.mpcd.types = test_types
+        snap.mpcd.mass = test_mass
+    sim = simulation_factory(snap, (1, 1, 2))
+
+    # reap all of the particle data into checkable list sorted by tag
+    pdata = sim._state._cpp_sys_def.getMPCDParticleData()
+    dat = []
+    for i in range(0, pdata.N):
+        pos_i = pdata.getPosition(i)
+        vel_i = pdata.getVelocity(i)
+        type_i = pdata.getType(i)
+        tag_i = pdata.getTag(i)
+        dat += [[
+            tag_i, pos_i.x, pos_i.y, pos_i.z, vel_i.x, vel_i.y, vel_i.z, type_i
+        ]]
+    dat = np.array(sorted(dat, key=lambda p: p[0]))
+
+    # now we do a per-processor check
+    if snap.communicator.num_ranks > 1:
+        if snap.communicator.rank == 0:
+            assert dat[0, 0] == 1
+            np.testing.assert_array_equal(dat[0, 1:4], test_positions[1])
+            np.testing.assert_array_equal(dat[0, 4:7], test_velocities[1])
+            assert dat[0, 7] == test_typeids[1]
+        else:
+            np.testing.assert_array_equal(dat[:, 0], [0, 2])
+            np.testing.assert_array_equal(dat[:, 1:4], test_positions[0::2])
+            np.testing.assert_array_equal(dat[:, 4:7], test_velocities[0::2])
+            np.testing.assert_array_equal(dat[:, 7], test_typeids[0::2])
+    else:
+        np.testing.assert_array_equal(dat[:, 0], [0, 1, 2])
+        np.testing.assert_array_equal(dat[:, 1:4], test_positions)
+        np.testing.assert_array_equal(dat[:, 4:7], test_velocities)
+        np.testing.assert_array_equal(dat[:, 7], test_typeids)
+
+
+def test_take_restore(snap, simulation_factory):
```

Comment:
There is a lot of repeat code between this test and the previous one, and they do very similar things. Can the code be reorganized so there is no duplicated code between them?